### ISSUE TITLE
nrf/boards/microbit: Backport persistent filesystem module

### ISSUE THREAD

**Issue by glennrub** [tralamazza/micropython#95]
Depends on hal/hal_nvmc to be implemented (issue #96).

Could driver be put under nrf/boards/microbit/extra_drivers or similar and have a new boards drivers hook to be compiled in?

Reference to micro:bit source: https://github.com/bbcmicrobit/micropython/blob/master/source/microbit/filesystem.c
Doc: http://microbit-micropython.readthedocs.io/en/latest/filesystem.html

**Comment by aykevl** [comment]
How should this be implemented?
I suspect it needs to be implemented using the VFS layer. But the underlying system can be done in two ways: via a block device and by directly talking to the HAL. The former is more flexible, the latter probably uses less code.
It could also be implemented in the MicroPython core, not just in our port. That way, other ports may use the same filesystem.
On the micro:bit it doesn't have a VFS layer and calls the underlying flash API directly.

**Comment by glennrub** [comment]
This task was originally intended to replicate the same microbit API, being a direct user of hal_nvmc when available. The idea was as well to expose the backported module to other targets (nrf51 and nrf52). But, you have a point when it comes to VFS.

The flash footprint of using raw hal_nvmc access will be significantly less then adding in FAT/VFS. So this is one of the targets where it makes sense to strip out FAT/VFS, as the original port did not have this. However, to also make microbit as a device a bit more attractive and updated on API (machine API, BLE support, FAT/VFS etc.) i have been thinking that this target (microbit) should probably have one more compile flag, LEGACY=1, to distinguish between legacy mode and modern API. So i would say this backport task belongs to the legacy mode. However, the tiny flash access module could potentially be reused by other targets.

As you understand, i have not thought this through fully. I'm very open to suggestions here.

**Comment by aykevl** [comment]
(I updated my comment above before I saw your comment)

I did a test and it turns out FAT requires somewhere between 20 and 50kB of storage. With a SoftDevice, there is only 10kB. So using FAT for on-chip storage is impossible on the nRF51+SoftDevice.

I see a few options (more combinations possible):
* keep VFS, blockdev and FAT but add the microbit filesystem directly talking to hal_nvmc. This may be useful for systems that might add SD card support but also want to use what's left of internal storage.
* drop both VFS and blockdev, and re-implement `open`, `os.listdir`, `os.remove`, etc - OR implement the micro:bit API.

I think it may be best to port the microbit filesystem over to hal_nvmc (without FAT/VFS), and if it makes sense maybe integrate it with VFS after that. Each board may then be configured to either use an SD card only (with FAT), use internal storage only (with microbit filesystem), or both. I'm hesitant about adding SD card support by default (and thus FAT) when a board doesn't have an SD card adapter.
